File: apps/web/src/services/api-client.ts
import axios, { type AxiosError, type InternalAxiosRequestConfig } from "axios";
import { useAuthStore } from "@/stores/auth-store";

const rawBaseURL = import.meta.env.VITE_API_BASE?.trim();
export const API_BASE_URL = rawBaseURL && rawBaseURL.length > 0 ? rawBaseURL : "/api/v1";

export const api = axios.create({ baseURL: API_BASE_URL });

if (import.meta.env.DEV) {
  console.info("[api] Base URL resolved to:", API_BASE_URL);
}

api.interceptors.request.use((config: InternalAxiosRequestConfig) => {
  const token = useAuthStore.getState().tokens?.accessToken;
  if (token) config.headers.Authorization = `Bearer ${token}`;
  return config;
});

let refreshPromise: Promise<string | null> | null = null;

api.interceptors.response.use(
  (r) => r,
  async (error: AxiosError) => {
    const original = error.config as InternalAxiosRequestConfig & { _retry?: boolean };
    const refreshToken = useAuthStore.getState().tokens?.refreshToken;

    if (error.response?.status === 401 && refreshToken && !original._retry) {
      original._retry = true;
      refreshPromise ??= refreshAccessToken(refreshToken).finally(() => {
        refreshPromise = null;
      });
      const newAccess = await refreshPromise;
      if (newAccess) {
        original.headers.Authorization = `Bearer ${newAccess}`;
        return api(original);
      }
      useAuthStore.getState().clear();
    }

    return Promise.reject(error);
  },
);

async function refreshAccessToken(refreshToken: string): Promise<string | null> {
  try {
    const res = await axios.post(`${API_BASE_URL}/auth/refresh`, { refreshToken });
    const tokens = res.data?.data;
    if (tokens) {
      useAuthStore.getState().setTokens(tokens);
      return tokens.accessToken as string;
    }
  } catch {
    // fall through
  }
  return null;
}

File: apps/flutter_mobile/lib/core/network/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter_mobile/core/auth/auth_events.dart';
import 'package:flutter_mobile/core/config/app_config.dart';
import 'package:flutter_mobile/core/constants/api_routes.dart';
import 'package:flutter_mobile/core/errors/api_exception.dart';
import 'package:flutter_mobile/core/network/api_envelope.dart';
import 'package:flutter_mobile/core/storage/secure_token_store.dart';
import 'package:flutter_mobile/features/auth/data/auth_models.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

final apiClientProvider = Provider<ApiClient>((ref) {
  return ApiClient(
    const SecureTokenStore(),
    onAuthExpired: () {
      ref.read(authExpiredSignalProvider.notifier).notifyExpired();
    },
  );
});

class ApiClient {

  ApiClient(this._tokenStore, {this.onAuthExpired}) {
    const baseUrl = AppConfig.apiBaseUrl;

    _dio = Dio(BaseOptions(
      baseUrl: baseUrl,
      connectTimeout: const Duration(seconds: 120),
      receiveTimeout: const Duration(seconds: 120),
    ));

    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        final tokens = await _tokenStore.read();
        if (tokens != null && tokens.accessToken.isNotEmpty) {
          options.headers['Authorization'] = 'Bearer ${tokens.accessToken}';
        }
        return handler.next(options);
      },
      onError: (e, handler) async {
        // Handle 401 Unauthorized
        if (e.response?.statusCode == 401 && !e.requestOptions.path.contains('/auth/refresh')) {
          final tokens = await _tokenStore.read();
          if (tokens != null && tokens.refreshToken.isNotEmpty) {
            try {
              final newAccessToken = await _refreshAccessToken(tokens.refreshToken);
              if (newAccessToken != null) {
                // Retry original request
                final opts = Options(
                  method: e.requestOptions.method,
                  headers: e.requestOptions.headers,
                );
                opts.headers?['Authorization'] = 'Bearer $newAccessToken';
                final cloneReq = await _dio.request<dynamic>(
                  e.requestOptions.path,
                  options: opts,
                  data: e.requestOptions.data,
                  queryParameters: e.requestOptions.queryParameters,
                );
                return handler.resolve(cloneReq);
              }
            } on Exception {
              await _clearAuthSession();
            }
          }
        }

        // Parse backend errors into ApiException
        ApiErrorDetail? apiError;
        if (e.response?.data != null && e.response?.data is Map<String, dynamic>) {
          try {
            final data = e.response!.data as Map<String, dynamic>;
            if (data.containsKey('error')) {
              apiError = ApiErrorDetail.fromJson(data['error'] as Map<String, dynamic>);
            }
          } on Object catch (_) {
            // Ignore parse errors
          }
        }

        return handler.next(
          DioException(
            requestOptions: e.requestOptions,
            response: e.response,
            type: e.type,
            error: ApiException(
              message: apiError?.message ?? e.message ?? 'Unknown error',
              errorDetail: apiError,
              statusCode: e.response?.statusCode,
            ),
          ),
        );
      },
    ));
  }
  late final Dio _dio;
  final SecureTokenStore _tokenStore;
  final void Function()? onAuthExpired;
  Future<String?>? _refreshFuture;
  bool _hasNotifiedAuthExpired = false;

  Dio get dio => _dio;

  Future<String?> _refreshAccessToken(String refreshToken) async {
    if (_refreshFuture != null) {
      return _refreshFuture;
    }

    return _refreshFuture = () async {
      try {
        final response = await _dio.post<Map<String, dynamic>>(ApiRoutes.authRefresh, data: {
          'refreshToken': refreshToken,
        });
        
        final dataMap = (response.data ?? {})['data'] as Map<String, dynamic>;
        final tokens = AuthTokens.fromJson(dataMap);
        
        await _tokenStore.save(tokens);
        _hasNotifiedAuthExpired = false;
        return tokens.accessToken;
      } on Exception {
        rethrow;
      } finally {
        _refreshFuture = null;
      }
    }();
  }

  Future<void> _clearAuthSession() async {
    await _tokenStore.delete();
    if (_hasNotifiedAuthExpired) return;
    _hasNotifiedAuthExpired = true;
    onAuthExpired?.call();
  }
}
